lib/http/header_normalizer.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+module HTTP
+  class HeaderNormalizer
+    # Matches HTTP header names when in "Canonical-Http-Format"
+    CANONICAL_NAME_RE = /\A[A-Z][a-z]*(?:-[A-Z][a-z]*)*\z/
+
+    # Matches valid header field name according to RFC.
+    # @see http://tools.ietf.org/html/rfc7230#section-3.2
+    COMPLIANT_NAME_RE = /\A[A-Za-z0-9!#$%&'*+\-.^_`|~]+\z/
+
+    MAX_CACHE_SIZE = 200
+
+    def initialize
+      @cache = LRUCache.new(MAX_CACHE_SIZE)
+    end
+
+    # Transforms `name` to canonical HTTP header capitalization
+    def normalize(name)
+      @cache[name] ||= normalize_header(name)
+    end
+
+    private
+
+    # Transforms `name` to canonical HTTP header capitalization
+    #
+    # @param [String] name
+    # @raise [HeaderError] if normalized name does not
+    #   match {COMPLIANT_NAME_RE}
+    # @return [String] canonical HTTP header name
+    def normalize_header(name)
+      return name if CANONICAL_NAME_RE.match?(name)
+
+      normalized = name.split(/[\-_]/).each(&:capitalize!).join("-")
+
+      return normalized if COMPLIANT_NAME_RE.match?(normalized)
+
+      raise HeaderError, "Invalid HTTP header field name: #{name.inspect}"
+    end
+
+    class LRUCache
+      def initialize(max_size)
+        @max_size = max_size
+        @cache = {}
+        @order = []
```

Comment:
Ruby hashes are already ordered in terms of insertion order. You don't need to maintain a separate data structure to preserve the order.

Reply:
I pushed a commit; let me know what you think. The change I pushed is no longer an LRU cache; it's a cache where the oldest item based on insertion order gets removed when the cache overflows. It's therefore a bit simpler - maybe less efficient in some cases but it certainly works well for the case I'm optimizing for.

Reply:
That looks great! Was there any impact on performance?

Reply:
I guess to benchmark against an LRU you'd need something simulating an actual workload. An LRU may be a better tool for the job.

Reply:
The change makes it about 35% faster on my Intel-based MacBook! I just refreshed the benchmarks in the description, using this laptop. (Yesterday I was on my work computer which has M3 chip and all the numbers were faster, but I refreshed all the benchmarks to be apples-to-apples, no pun intended.)

Reply:
> I guess to benchmark against an LRU you'd need something simulating an actual workload. An LRU may be a better tool for the job.

Yeah I do agree with that. I am happy to revert the LRUCache -> Cache commit if you think it was better before. I can also tweak the implementation if you'd prefer I not use the array and instead use the Hash to maintain ordering and do something like delete the element from the hash and re-add it.

Reply:
If the initial benchmarks are promising, it would also be nice to have something lightweight to start with, and maybe circle back on an LRU if your real-world telemetry shows otherwise

Reply:
Sounds good! I suspect that my system will not need LRU because it sees the same headers over and over.